### PR TITLE
Better Rust API (tch-rs)

### PR DESCRIPTION
**Pull request by Narsil**

### CODE REVIEW

File: safetensors/src/tensor.rs [hunk 1]
```diff
@@ -48,34 +49,126 @@ impl std::fmt::Display for SafeTensorError {
...
     // Then by name
-    let mut data: Vec<_> = data.iter().collect();
+    let mut data: Vec<_> = data.into_iter().collect();
     data.sort_by(|(lname, left), (rname, right)| {
-        right.dtype.cmp(&left.dtype).then(lname.cmp(rname))
+        right.dtype().cmp(&left.dtype()).then(lname.cmp(rname))
     });
 
-    let mut tensors: Vec<&TensorView> = vec![];
+    let mut tensors: Vec<V> = vec![];
```
Comment (co42) [inline safetensors/src/tensor.rs:163]: You can do `Vec::with_capacity` like below

File: safetensors/src/tensor.rs [hunk 2]
```diff
@@ -48,34 +49,126 @@ impl std::fmt::Display for SafeTensorError {
 
 impl std::error::Error for SafeTensorError {}
 
-struct PreparedData<'hash, 'data> {
+struct PreparedData {
     n: u64,
     header_bytes: Vec<u8>,
-    tensors: Vec<&'hash TensorView<'data>>,
+    // tensors: Vec<&'data TensorView<'data>>,
```
Comment (co42) [inline safetensors/src/tensor.rs:55]: Debug ?

File: safetensors/src/tensor.rs [hunk 3]
```diff
@@ -48,34 +49,126 @@ impl std::fmt::Display for SafeTensorError {
...
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        // This copies data from GPU to CPU.
+///        let data: Vec<u8> = self.data.to_vec();
+///        data.into()
+///    }
+///    fn data_len(&self) -> usize{
```
Comment (co42) [inline safetensors/src/tensor.rs:131]:
```suggestion
/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: Vec<u8> }
///
/// // type MyDtype = Dtype;
/// impl<'data> View for &'data Tensor {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        (&self.data).into()
///    }
///    fn data_len(&self) -> usize {
///        self.data.len()
///    }
/// }
/// ```
///
/// For a borrowed tensor:
///
/// ```rust
/// use safetensors::tensor::{View, Dtype};
/// use std::borrow::Cow;
/// struct Tensor<'data>{ dtype: MyDtype, shape: Vec<usize>, data: &'data[u8] }
///
/// // type MyDtype = Dtype;
/// impl<'data> View for Tensor<'data> {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        self.data.into()
///    }
///    fn data_len(&self) -> usize {
///        self.data.len()
///    }
/// }
/// ```
///
/// Now if you have some unknown buffer that could be on GPU for instance,
/// you can implement the trait to return an owned local buffer containing the data
/// on CPU (needed to write on disk)
/// ```rust
/// use safetensors::tensor::{View, Dtype};
/// use std::borrow::Cow;
///
/// // type MyDtype = Dtype;
/// // type OpaqueGpu = Vec<u8>;
/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: OpaqueGpu }
///
/// impl View for Tensor {
///    fn dtype(&self) -> Dtype {
///        self.dtype.into()
///    }
///    fn shape(&self) -> &[usize] {
///         &self.shape
///    }
///    fn data(&self) -> Cow<[u8]> {
///        // This copies data from GPU to CPU.
///        let data: Vec<u8> = self.data.to_vec();
///        data.into()
///    }
///    fn data_len(&self) -> usize {
```

**Review by co42 — APPROVED** [review]

File: safetensors/src/tensor.rs (hunk 1 above)
Comment (Narsil) [reply safetensors/src/tensor.rs:163]: Thanks.

File: safetensors/src/tensor.rs (hunk 2 above)
Comment (Narsil) [reply safetensors/src/tensor.rs:55]: Yup

**Pull request merged**